src/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -1,228 +1,178 @@
         }
 
-        public ExtensionEntry LoadExtension(ExtensionDescriptor extensionDescriptor)
+        public Task<ExtensionEntry> LoadExtensionAsync(IExtensionInfo extensionInfo)
         {
             // Results are cached so that there is no mismatch when loading an assembly twice.
             // Otherwise the same types would not match.
-
-            try
+            return _extensions.GetOrAdd(extensionInfo.Id, async id =>
             {
-                return _extensions.GetOrAdd(extensionDescriptor.Id, id =>
-                {
-                    foreach (var loader in _loaders)
-                    {
-                        ExtensionEntry entry = loader.Load(extensionDescriptor);
-                        if (entry != null)
-                        {
-                            return entry;
-                        }
-                    }
+                var extension = _extensionLoader.Load(extensionInfo);
 
-                    if (_logger.IsEnabled(LogLevel.Warning))
-                    {
-                        _logger.LogWarning("No suitable loader found for extension \"{0}\"", extensionDescriptor.Id);
-                    }
+                if (extension.IsError && _logger.IsEnabled(LogLevel.Warning))
+                {
+                    _logger.LogWarning("No suitable loader found for extension \"{0}\"", id);
+                }
 
-                    return null;
-                });
-            }
-            catch (Exception ex)
-            {
-                _logger.LogError(string.Format("Error loading extension '{0}'", extensionDescriptor.Id), ex);
-                throw new OrchardException(T("Error while loading extension '{0}'.", extensionDescriptor.Id), ex);
-            }
+                return await Task.FromResult(extension);
+            });
         }
 
-        public IEnumerable<Feature> LoadFeatures(IEnumerable<FeatureDescriptor> featureDescriptors)
-        {
-            if (_logger.IsEnabled(LogLevel.Information))
-            {
-                _logger.LogInformation("Loading features");
-            }
-
-            var result = featureDescriptors
-                .Select(descriptor => LoadFeature(descriptor))
-                .ToArray();
+        public async Task<IEnumerable<ExtensionEntry>> LoadExtensionsAsync(IEnumerable<IExtensionInfo> extensionInfos) {
+            var entries = new List<ExtensionEntry>();
 
-            if (_logger.IsEnabled(LogLevel.Information))
-            {
-                _logger.LogInformation("Done loading features");
+            foreach (var extensionInfo in extensionInfos) {
+                var extension = await LoadExtensionAsync(extensionInfo);
+                entries.Add(extension);
             }
-            return result;
+
+            return entries;
         }
 
-        private Feature LoadFeature(FeatureDescriptor featureDescriptor)
+        public Task<FeatureEntry> LoadFeatureAsync(IFeatureInfo feature)
```

Comment:
We don't need both `LoadFeaturesAsync` and `LoadFeatureAsync`, the one with multiple features is just adding a foreach. And if we have only one, then we need the IEnumerable one only in case there can be optimizations.